use-cases/using-Intel-QAT-and-kata.md

```diff
@@ -0,0 +1,491 @@
+
+## Build OpenSSL Intel QAT engine container
+
+There is an OpenSSL Intel QAT Dockerfile located at 
+[https://github.com/intel/intel-device-plugins-for-kubernetes/tree/master/demo/openssl-qat-engine](https://github.com/intel/intel-device-plugins-for-kubernetes/tree/master/demo/openssl-qat-engine). 
+Use this file to build a container image with an optimized OpenSSL engine for 
+Intel QAT. Using Docker build with the kata runtime can sometimes have issues. 
+Therefore, it is recommended to change the default runtime to runc before 
+doing a build. Instructions for this are below.
+
+__Note:__ The Intel QAT driver version in this container might not match the 
+Intel QAT driver compiled and loaded on the host.
+
+```
+$ cd $QAT_SRC
+$ curl -O $QAT_DOCKERFILE
+$ sudo sed -i 's/kata-runtime/runc/g' /etc/systemd/system/docker.service.d/50-runtime.conf
+$ sudo systemctl daemon-reload && sudo systemctl restart docker
+$ docker build -t openssl-qat-engine .
+```
+
+### Test Intel QAT in Docker
+
+The host should already be setup with 16 virtual functions of the Intel QAT 
+card bound to VFIO-PCI. This can be verified by looking in `/dev/VFIO` for a 
+listing of functions. Replace the number 90 with one of the VF’s exposed 
+in `/dev/VFIO`. It might be required to add an `IPC_LOCK` capability to your 
+Docker runtime depending on which rootfs is used. 
+
+```
+$ docker run -it --runtime=kata-runtime --cap-add=IPC_LOCK --device=/dev/VFIO/90 -v /dev:/dev -v ${QAT_CONF_LOCATION}:/etc openssl-qat-engine bash
+```
+
+Below are some commands to run in the container image to verify Intel QAT is 
+working
+
+```
+bash-5.0# cat /proc/modules
+bash-5.0# adf_ctl restart
+bash-5.0# adf_ctl status
+bash-5.0# openssl engine -c -t qat
+
+#Test with Intel QAT card acceleration
+bash-5.0# openssl speed -engine qat -elapsed -async_jobs 72 rsa2048 
+
+#Test with CPU acceleration
+bash-5.0# openssl speed -elapsed rsa2048
+```
+
+### Troubleshooting
+
+* Check that `/dev/VFIO` has VF’s enabled
+```
+$ ls /dev/VFIO
+57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72  VFIO
+```
+* Check that the modules are loaded when inside the Kata Container
+```
+bash-5.0# cat /proc/modules
+qat_c62xvf 16384 - - Live 0x0000000000000000 (O)
+usdm_drv 86016 - - Live 0x0000000000000000 (O)
+intel_qat 184320 - - Live 0x0000000000000000 (O)
+```
+
+* Verify that at least the first `c6xxvf_dev0.conf` files is mounted inside the 
+container image in `/etc`. You will need one conf file for each VF passed into 
+the container.
+```
+bash-5.0# ls /etc
+c6xxvf_dev0.conf   c6xxvf_dev11.conf  c6xxvf_dev14.conf  c6xxvf_dev3.conf  c6xxvf_dev6.conf  c6xxvf_dev9.conf  resolv.conf
+c6xxvf_dev1.conf   c6xxvf_dev12.conf  c6xxvf_dev15.conf  c6xxvf_dev4.conf  c6xxvf_dev7.conf  hostname
+c6xxvf_dev10.conf  c6xxvf_dev13.conf  c6xxvf_dev2.conf   c6xxvf_dev5.conf c6xxvf_dev8.conf  hosts
+```
+
+* Check dmesg inside the container to see if there are any issues with the 
```

Comment:
`dmesg`.

Reply:
Updated

Reply:
Hi @eadamsintel - did you forget to re-push? I'm still seeing the old version of the file.

Reply:
Sorry.  I was writing update in here as I updated so I don't lose track of all the recommendations. I thought I would have time to go through the entire instructions again to verify they still work but the rootfs download took hours and hours. I should be able to resubmit today. I'll let you and Graham know when I do.